keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,638 @@
+-->
+
+- `<package>`: `<date>` - `<test coverage>`
+
+##### Integration tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- <test>: <link to test coverage>
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+- <test>: <link to test coverage>
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Complete benchmarks
+
+#### GA
+
+- Add metrics
+- Add a full documentation with examples for the most popular providers, e.g., Okta, Dex, Auth0
+- Migration guide
+- Deprecation warnings for non-stuctured OIDC provider configuration
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+kube-apiserver `--oidc-*` flags require deprecation warnings on the stable release of the feature.
+It is possible to react only to the `--oidc-issuer-url` flag because other flags cannot be enabled separately from this one.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- Feature gate
+  - Feature gate name: `StructuredOIDCConfiguration`
+  - Components depending on the feature gate:
+    - kube-apiserver
+
+```go
+FeatureSpec{
+	Default: false,
+	LockToDefault: false,
+	PreRelease: featuregate.Alpha,
+}
+```
+
+###### Does enabling the feature change any default behavior?
+
+No.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, but disable means also deleting the flag from the kube-apiserver manifest.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+No impact.
+
+###### Are there any tests for feature enablement/disablement?
+
+Not required.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+It cannot fail.
```

Comment:
Some possible rollout/rollback failures I can imagine are:
* General bugs in kube-apiserver's parsing of the structured config file. Maybe more relevant in future versions where there could be bugs in the conversion logic between multiple versions (v1alpha1, v1beta1, etc) of the config.  
* Cluster admin rolls out the feature with the addition of some validation rules that may have some untended consequences
* Rolling back would mean losing validation functionality offered from structured config that some other component may depend on

(not suggesting much can be done about all of these, especially if they are due to user error, but probably worth noting here)

Reply:
Yes. And noting that failures are intended to prevent process startup means that HA clusters that ripple-start will survive intact.